### PR TITLE
fix(developer): raise error if virtual key in context string

### LINKED ISSUE #7880: bug(developer): Keyman Developer does not flag virtual keycodes in context string as an error
### Describe the bug

If you use a virtual key code in the context part of a rule, Keyman Developer does not tell you this is an error. But the language guide says virtual key codes can only work in the keystroke part of the rule. A rule with a virtual keycode in the context will not work. 

See this screenshot. The rule will never work, but the Developer compiles it without reporting a problem.

image

### Reproduce the bug

_No response_

### Expected behavior

_No response_

### Related issues

_No response_

### Keyman apps

- [ ] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [X] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

_No response_

### Operating system

Windows 11 22H2

### Device

Windows laptop

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #7880.

Virtual keys have never been allowed in context. This should be an error in a keyboard. Note: virtual keys in output are officially unsupported but still kinda sorta a little bit work in Windows -- but I doubt they will ever be officially supported.

Also includes small side journey to tidy up names and constants for two other tests (error_duplicate_group and error_duplicate_store).

@keymanapp-test-bot skip

### CODE REVIEW

File: developer/src/kmc-kmn/test/test-messages.ts [hunk 1]
```diff
@@ -54,15 +54,22 @@ describe('CompilerMessages', function () {
   // CERR_DuplicateGroup
 
   it('should generate CERR_DuplicateGroup if the kmn contains two groups with the same name', async function() {
-    await testForMessage(this, ['invalid-keyboards', 'cerr_duplicate_group.kmn'], 0x302071); //TODO: consolidate messages from kmcmplib, CompilerMessages.CERR_DuplicateGroup
+    await testForMessage(this, ['invalid-keyboards', 'error_duplicate_group.kmn'], KmnCompilerMessages.ERROR_DuplicateGroup); //TODO: consolidate messages from kmcmplib, CompilerMessages.CERR_DuplicateGroup
```
Comment (mcdurdin) [inline developer/src/kmc-kmn/test/test-messages.ts:57]:
```suggestion
    await testForMessage(this, ['invalid-keyboards', 'error_duplicate_group.kmn'], KmnCompilerMessages.ERROR_DuplicateGroup);
```

File: developer/src/kmc-kmn/test/fixtures/invalid-keyboards/error_virtual_key_in_context.kmn [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+begin Unicode > use(main)
+
+group(main) using keys
+store (consonant) "lmn"
+store (sylcons) U+1E37 U+1E43 U+1E47
+
+
+c ERROR_VirtualKeyInContext
+[K_BKQUOTE] + any(consonant) > index (sylcons, 2)
```
Comment (mcdurdin) [inline developer/src/kmc-kmn/test/fixtures/invalid-keyboards/error_virtual_key_in_context.kmn:12]:
```suggestion
store (consonant) "lmn"
store (sylcons) U+1E37 U+1E43 U+1E47


c ERROR_VirtualKeyInContext
[K_BKQUOTE] + any(consonant) > index (sylcons, 2)
```
```suggestion

c ERROR_VirtualKeyInContext
[K_BKQUOTE] + 'a' > 'b'
```

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 3]
```diff
@@ -1316,6 +1316,9 @@ KMX_BOOL CheckContextStatementPositions(PKMX_WCHAR context) {
           AddWarningBool(CWARN_IfShouldBeAtStartOfContext);
         }
         break;
+      case CODE_EXTENDED:
+        AddCompileError(CERR_VirtualKeyInContext);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:1320]: Update the function comment to note this test.

**Review by rc-swag — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.204-alpha](https://keyman.com/downloads/releases/alpha/17.0.204)

**Pull request merged**